diff -ruN linux-2.4.19-rc1/drivers/md/Config.in linux/drivers/md/Config.in
--- linux-2.4.19-rc1/drivers/md/Config.in	Fri Sep 14 22:22:18 2001
+++ linux/drivers/md/Config.in	Wed Jan  2 19:23:58 2002
@@ -14,5 +14,8 @@
 dep_tristate '  Multipath I/O support' CONFIG_MD_MULTIPATH $CONFIG_BLK_DEV_MD
 
 dep_tristate ' Logical volume manager (LVM) support' CONFIG_BLK_DEV_LVM $CONFIG_MD
+if [ "$CONFIG_EXPERIMENTAL" = "y" ]; then
+   dep_tristate ' Device-mapper support (EXPERIMENTAL)' CONFIG_BLK_DEV_DM $CONFIG_MD
+fi
 
 endmenu
diff -ruN a/Documentation/Configure.help b/Documentation/Configure.help
--- linux-2.4.19-rc1/Documentation/Configure.help	Tue Jun 25 14:14:05 2002
+++ linux/Documentation/Configure.help	Tue Jun 25 19:18:26 2002
@@ -1775,6 +1775,18 @@
   want), say M here and read <file:Documentation/modules.txt>.  The
   module will be called lvm-mod.o.
 
+Device-mapper support
+CONFIG_BLK_DEV_DM
+  This option lets you create logical block devices dynamically by
+  joining together segments of existing block devices.  This mechanism
+  is used by the new version of the logical volume manager under
+  development, LVM2.
+
+  If you want to compile this as a module, say M here and read 
+  <file:Documentation/modules.txt>.  The module will be called dm-mod.o.
+
+  If unsure, say N.
+
 Multiple devices driver support (RAID and LVM)
 CONFIG_MD
   Support multiple physical spindles through a single logical device.
