diff -ruN linux-2.4.19-rc1/include/linux/fs.h linux/include/linux/fs.h
--- linux-2.4.19-rc1/include/linux/fs.h	Tue Feb 19 15:24:57 2002
+++ linux/include/linux/fs.h	Thu Feb 21 12:34:42 2002
@@ -260,7 +260,10 @@
 	char * b_data;			/* pointer to data block */
 	struct page *b_page;		/* the page this bh is mapped to */
 	void (*b_end_io)(struct buffer_head *bh, int uptodate); /* I/O completion */
- 	void *b_private;		/* reserved for b_end_io */
+ 	void *b_private;		/* reserved for b_end_io, also used by ext3 */
+	void *b_bdev_private;	        /* a hack to get around ext3 using b_private
+					 * after handing the buffer_head to the
+					 * block layer */
 
 	unsigned long b_rsector;	/* Real buffer location on disk */
 	wait_queue_head_t b_wait;
