diff -Nru a/mm/vmalloc.c b/mm/vmalloc.c
--- a/mm/vmalloc.c	Wed Jun 12 12:04:44 2002
+++ b/mm/vmalloc.c	Thu Jun 13 13:13:44 2002
@@ -321,3 +321,22 @@
 	read_unlock(&vmlist_lock);
 	return buf - buf_start;
 }
+
+void *vcalloc(unsigned long nmemb, unsigned long size)
+{
+	unsigned long len;
+	void *mem;
+
+	/*
+	 * Check that we're not going to overflow.
+	 */
+	if (nmemb > (ULONG_MAX / size))
+		return NULL;
+
+	len = nmemb * size;
+	mem = vmalloc(len);
+	if (mem)
+		memset(mem, 0, len);
+
+	return mem;
+}
diff -Nru a/include/linux/vmalloc.h b/include/linux/vmalloc.h
--- a/include/linux/vmalloc.h	Wed Jun 12 12:35:58 2002
+++ b/include/linux/vmalloc.h	Thu Jun 13 13:13:39 2002
@@ -25,6 +25,7 @@
 extern void vmfree_area_pages(unsigned long address, unsigned long size);
 extern int vmalloc_area_pages(unsigned long address, unsigned long size,
                               int gfp_mask, pgprot_t prot);
+extern void *vcalloc(unsigned long nmemb, unsigned long size);
 
 /*
  *	Allocate any pages
