diff -ruN linux-2.4.17/fs/namespace.c linux/fs/namespace.c
--- linux-2.4.17/fs/namespace.c	Wed Jan  2 19:10:48 2002
+++ linux/fs/namespace.c	Wed Jan  2 19:32:35 2002
@@ -289,7 +289,7 @@
 	}
 }
 
-static int do_umount(struct vfsmount *mnt, int flags)
+int do_umount(struct vfsmount *mnt, int flags)
 {
 	struct super_block * sb = mnt->mnt_sb;
 	int retval = 0;
diff -ruN linux-2.4.17/include/linux/fs.h linux/include/linux/fs.h
--- linux-2.4.17/include/linux/fs.h	Wed Jan  2 19:10:48 2002
+++ linux/include/linux/fs.h	Wed Jan  2 19:32:35 2002
@@ -983,6 +983,7 @@
 extern struct vfsmount *kern_mount(struct file_system_type *);
 extern int may_umount(struct vfsmount *);
 extern long do_mount(char *, char *, char *, unsigned long, void *);
+extern int do_umount(struct vfsmount *mnt, int flags);
 
 #define kern_umount mntput
 
